### PR TITLE
Grad clip for parameters on different devices

### PR DESCRIPTION
**Pull request by Stonesjtu**
I'm trying to write a multi-gpu network by pipelining some layers onto different GPUs. However, the current gradient clip requires all the parameters to locate in the same device.

The overhead of CUDA launch is reduced since the scalar calculation is performed on CPU, but it introduces extra data transfers.

No performance regression is observed by running the following snippet:
```python
import time

import torch

module = torch.nn.Sequential(
    torch.nn.LSTM(1024, 1024),
    torch.nn.LSTM(256, 256),
    torch.nn.Linear(100, 10000),
).cuda()

# warming-up
torch.nn.utils.clip_grad_norm_(module.parameters(), 1)
torch.cuda.synchronize()
start = time.time()
for _ in range(1000):
    torch.nn.utils.clip_grad_norm_(module.parameters(), 1)
torch.cuda.synchronize()
time_elapse = time.time() - start
print('{} ms per clip'.format(time_elapse))
```

### CODE REVIEW

**Review by soumith — APPROVED** [review]

**Comment by soumith** [comment]
thank you, this looks good!

**Comment by apaszke** [comment]
Wouldn’t it be much faster to just use a defaultdict to accumulate norms on different devices and only them transfer them all to CPU? That would at least cover the common case of all params on a single GPU

**Comment by Stonesjtu** [comment]
@apaszke I've thought about that, but counter-intuitively on my environment, the scalar addition on a single device does not run as fast as expected.

My envs:
1080 + Xeon E5 2620 V4.
